### PR TITLE
Fix Web and Native build

### PR DESCRIPTION
**Pull request by shrekshao**
Modify on top of @munrocket's pull request: https://github.com/kainino0x/webgpu-cross-platform-demo/pull/8

* Update dawn version in `setup_native_build.sh`
* Fix native, verified build and run on linux
* For native, select adapter with some simple heuristic (GPU > CPU, Modern Graphics API > Older API)

### CODE REVIEW

**Comment by munrocket** [comment]
super!

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,10 +29,14 @@ if(EMSCRIPTEN)
         #PRIVATE "SHELL:-s VERBOSE=1"
         )
 else()
+    add_compile_definitions(DAWN_ENABLE_DESKTOP_GL=OFF)
```
Comment (shrekshao) [inline CMakeLists.txt:32]: Looks like these won't actually passed down to dawn added via `add_subdirectory`. Maybe we should modify dawn CMakeList.txt to use PARENT_SCOPE for the options.
But this is lower priroity
https://levelup.gitconnected.com/cmake-variable-scope-f062833581b7

Right now we can use `cmake ../.. -D DAWN_ENABLE_DESKTOP_GL=OFF DAWN_ENABLE_OPENGLES=OFF` to disable them

Comment (munrocket) [reply CMakeLists.txt:32]: I can't build with actual dawn revision on macOS too and started to compile with 4b01515.

Comment (munrocket) [reply CMakeLists.txt:32]: >cmake ../.. -D DAWN_ENABLE_DESKTOP_GL=OFF DAWN_ENABLE_OPENGLES=OFF

This is easier to understand and fix later

Comment (shrekshao) [reply CMakeLists.txt:32]: Does your error looks like something like this?
```
webgpu-cross-platform-demo/third_party/dawn/src/dawn/native/metal/CommandRecordingContext.h:40:41: error: 
      unknown type name 'MTLBlitPassDescriptor'; did you mean 'MTLRenderPassDescriptor'?
    id<MTLBlitCommandEncoder> BeginBlit(MTLBlitPassDescriptor* descriptor)
                                        ^~~~~~~~~~~~~~~~~~~~~
                                        MTLRenderPassDescriptor
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Metal.framework/Headers/MTLCommandBuffer.h:22:8: note: 
      'MTLRenderPassDescriptor' declared here
@class MTLRenderPassDescriptor;
```

Looks like some metal configuration issue...

Comment (shrekshao) [reply CMakeLists.txt:32]: Hmm https://developer.apple.com/documentation/metal/mtlblitpassdescriptor

Maybe the dawn src needs to bracket MTLBlitPassDescriptor with
`if (@available(macOS 11.0, iOS 14.0, *))`?

Comment (shrekshao) [reply CMakeLists.txt:32]: Or maybe I need to update my MacOS sdk

Comment (munrocket) [reply CMakeLists.txt:32]: On new MacOS 12.6 and xcode, with additional flags to turn off OpenGL error looks like this:

```sh
[ 95%] Building CXX object third_party/dawn/src/dawn/native/CMakeFiles/dawn_native.dir/metal/BackendMTL.mm.o
/Users/munrocket/code/git/webgpu-cross-platform-demo/third_party/dawn/src/dawn/native/metal/BackendMTL.mm:124:50: warning: 'kIOMasterPortDefault' is deprecated: first deprecated in macOS 12.0 [-Wdeprecated-declarations]
        AcquireIORef(IOServiceGetMatchingService(kIOMasterPortDefault, matchingDict.Detach()));
                                                 ^~~~~~~~~~~~~~~~~~~~
                                                 kIOMainPortDefault
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/System/Library/Frameworks/IOKit.framework/Headers/IOKitLib.h:123:19: note: 'kIOMasterPortDefault' has been explicitly marked deprecated here
const mach_port_t kIOMasterPortDefault
                  ^
1 warning generated.
[100%] Building CXX object CMakeFiles/hello.dir/main.cpp.o
In file included from /Users/munrocket/code/git/webgpu-cross-platform-demo/main.cpp:62:
/Users/munrocket/code/git/webgpu-cross-platform-demo/third_party/dawn/include/dawn/native/DawnNative.h:25:15: warning: nested namespace definition is a C++17 extension; define each namespace separately [-Wc++17-extensions]
namespace dawn::platform {
              ^~~~~~~~~~
               { namespace platform
/Users/munrocket/code/git/webgpu-cross-platform-demo/third_party/dawn/include/dawn/native/DawnNative.h:34:15: warning: nested namespace definition is a C++17 extension; define each namespace separately [-Wc++17-extensions]
namespace dawn::native {
              ^~~~~~~~
               { namespace native
/Users/munrocket/code/git/webgpu-cross-platform-demo/main.cpp:68:11: warning: enumeration value 'OpenGL' not handled in switch [-Wswitch]
  switch (t) {
          ^
/Users/munrocket/code/git/webgpu-cross-platform-demo/main.cpp:68:11: note: add missing switch cases
  switch (t) {
          ^
3 warnings generated.
[100%] Linking CXX executable hello
ld: unknown option: --copy-dt-needed-entries
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [hello] Error 1
make[1]: *** [CMakeFiles/hello.dir/all] Error 2
make: *** [all] Error 2
```

Comment (munrocket) [reply CMakeLists.txt:32]: Maybe this error related only for this relatively new macOS. I will check code later.

Comment (kainino0x) [reply CMakeLists.txt:32]: On my machine:
- `DAWN_ENABLE_DESKTOP_GL`/`DAWN_ENABLE_OPENGLES` default to OFF even without these lines
- I also get `ld: unknown option: --copy-dt-needed-entries` but the build works if I remove that flag
- I also get the deprecation warnings but they don't cause the build to fail
- I also get the C++ standard warning so I fixed that by upgrading the `CMAKE_CXX_STANDARD`

**Comment by munrocket** [comment]
Thanks!

File: CMakeLists.txt (hunk 1 above)
Comment (shrekshao) [reply CMakeLists.txt]: Turns out removing `--copy-dt-needed-entries` still works for me as well.

In a future update, we could follow these set DAWN macros to inject into dawn CMakeList scope
https://github.com/samdauwe/webgpu-native-examples/blob/master/CMakeLists.txt#L28

Comment (munrocket) [reply CMakeLists.txt]: Thanks for this link!

Comment (munrocket) [reply CMakeLists.txt]: @shrekshao native build is headless and only showing tests in console. Is there a way how to look at a picture?

Comment (shrekshao) [reply CMakeLists.txt]: I'm working on adding a window via glfw

Comment (munrocket) [reply CMakeLists.txt]: Ah ok. I didn't know that.

**Pull request merged**